alacritty/src/input.rs

```diff
@@ -1192,6 +1243,251 @@ impl<T: EventListener, A: ActionContext<T>> Processor<T, A> {
+        modifiers |= 0b0010;
+    }
+
+    if mods.control_key() {
+        modifiers |= 0b0100;
+    }
+
+    if mods.super_key() {
+        modifiers |= 0b1000;
+    }
+
+    // The `1` must be added to result.
+    modifiers += 1;
+
+    let named_csi_u = mode.intersects(
+        TermMode::KEYBOARD_REPORT_ALL_KEYS_AS_ESC
+            | TermMode::KEYBOARD_DISAMBIGUATE_ESC_CODES
+            | TermMode::KEYBOARD_REPORT_EVENT_TYPES,
+    );
+    // Send CSI u for numpad
+    let csi_u_numpad = key.location == KeyLocation::Numpad && named_csi_u;
+    let encode_all = mode.contains(TermMode::KEYBOARD_REPORT_ALL_KEYS_AS_ESC);
+    let send_event_type = mode.contains(TermMode::KEYBOARD_REPORT_EVENT_TYPES)
+        && (key.repeat || key.state == ElementState::Released);
+
+    let (codepoint, suffix): (Cow<'static, str>, char) = match key.logical_key.as_ref() {
+        // Special case numpad.
+        Key::Character("0") if csi_u_numpad => ("57399".into(), 'u'),
+        Key::Character("1") if csi_u_numpad => ("57400".into(), 'u'),
+        Key::Character("2") if csi_u_numpad => ("57401".into(), 'u'),
+        Key::Character("3") if csi_u_numpad => ("57402".into(), 'u'),
+        Key::Character("4") if csi_u_numpad => ("57403".into(), 'u'),
+        Key::Character("5") if csi_u_numpad => ("57404".into(), 'u'),
+        Key::Character("6") if csi_u_numpad => ("57405".into(), 'u'),
+        Key::Character("7") if csi_u_numpad => ("57406".into(), 'u'),
+        Key::Character("8") if csi_u_numpad => ("57407".into(), 'u'),
+        Key::Character("9") if csi_u_numpad => ("57408".into(), 'u'),
+        Key::Character(".") if csi_u_numpad => ("57409".into(), 'u'),
+        Key::Character("/") if csi_u_numpad => ("57410".into(), 'u'),
+        Key::Character("*") if csi_u_numpad => ("57411".into(), 'u'),
+        Key::Character("-") if csi_u_numpad => ("57412".into(), 'u'),
+        Key::Character("+") if csi_u_numpad => ("57413".into(), 'u'),
+        Key::Enter if csi_u_numpad => ("57414".into(), 'u'),
+        Key::Character("=") if csi_u_numpad => ("57415".into(), 'u'),
+        // KP_SEPARATOR if csi_u_numpad => ("57415".into(), 'u'),
+        Key::ArrowLeft if csi_u_numpad => ("57417".into(), 'u'),
+        Key::ArrowRight if csi_u_numpad => ("57418".into(), 'u'),
+        Key::ArrowUp if csi_u_numpad => ("57419".into(), 'u'),
+        Key::ArrowDown if csi_u_numpad => ("57420".into(), 'u'),
+        Key::PageUp if csi_u_numpad => ("57421".into(), 'u'),
+        Key::PageDown if csi_u_numpad => ("57422".into(), 'u'),
+        Key::Home if csi_u_numpad => ("57423".into(), 'u'),
+        Key::End if csi_u_numpad => ("57424".into(), 'u'),
+        Key::Insert if csi_u_numpad => ("57425".into(), 'u'),
+        Key::Delete if csi_u_numpad => ("57426".into(), 'u'),
+        // KP_BEGIN if csi_u_numpad => ("57427".into(), 'u'),
+        // Handle common keys.
+        Key::ArrowLeft if mods.is_empty() && !send_event_type => ("".into(), 'D'),
```

Comment:
This is a giant mess and I cannot imagine Rust manages to optimize all these conditionals well, especially when function calls are involved. Have you checked what kind of code this generates?